calyx/src/analysis/order_analysis.rs

```diff
@@ -0,0 +1,164 @@
+use crate::analysis::ReadWriteSet;
+use crate::ir;
+
+#[derive(Default)]
+///Primarily used to help determine the order cells are executed within
+///the group
+pub struct OrderAnalysis {
+    ///Holds, (b,a) for assignment of form a.go = b.done,
+    done_go: Option<(ir::Id, ir::Id)>,
+    ///Holds a for assignment of form group[done] = a.done
+    last: Option<ir::Id>,
+}
+
+impl OrderAnalysis {
+    //Returns whether the given assignment is a go done assignment from two cells of interest
+    //i.e. cell1.go = cell2.done.
+    pub fn is_go_done(asmt: &ir::Assignment) -> bool {
+        let src = asmt.src.borrow();
+        let dst = asmt.dst.borrow();
+        match (&src.parent, &dst.parent) {
+            (
+                ir::PortParent::Cell(src_cell),
+                ir::PortParent::Cell(dst_cell),
+            ) => {
+                //the first two checks may be unnecessary
+                Self::is_stateful(&src_cell.upgrade())
+                    && Self::is_stateful(&dst_cell.upgrade())
+                    && src.name == "done"
+                    && dst.attributes.has("go")
+            }
+            _ => false,
+        }
+    }
+    //Returns whether the given assignment writes to the go assignment of cell
+    //in the form cell.go = !cell.done? 1'd1.
+    pub fn is_specific_go(asmt: &ir::Assignment, cell: &ir::Id) -> bool {
+        //checks whether guard is cell.done
+        let guard_is_done = |guard: &ir::Guard| -> bool {
+            match guard {
+                ir::Guard::Port(port) => {
+                    port.borrow().name == "done"
```

Comment:
This is not the correct condition to check whether a port is a done port. You need to look for the `@done` attribute on a port definition

Reply:
For groups, it seems like checking for the `@done` attribute doesn't work.